### ISSUE TITLE
Double Softmax in DARC Implementation

### ISSUE THREAD

**Issue by stweigand97** [OffDynamicsRL/off-dynamics-rl#5]
Hi! I think that I might have spotted an issue with the DARC implementation of this work where the classifer's outputs are treated as logits, although a Softmax is applied already. In the computation of the reward penalty, for example, this leads to applying the Softmax function twice.

The relevant parts are the forward pass in the ``Classifier`` where the Softmax is applied:

`````python
    def forward(self, state_batch, action_batch, nextstate_batch, with_noise):
        sas = torch.cat([state_batch, action_batch, nextstate_batch], -1)

        if with_noise:
            sas += torch.randn_like(sas, device=state_batch.device) * self.gaussian_noise_std
        sas_logits = torch.nn.Softmax()(self.sas_classifier(sas))

        sa = torch.cat([state_batch, action_batch], -1)

        if with_noise:
            sa += torch.randn_like(sa, device=state_batch.device) * self.gaussian_noise_std
        sa_logits = torch.nn.Softmax()(self.sa_classifier(sa))

        return sas_logits, sa_logits
`````

and the section where the reward penalty is computed but the Softmax is applied again:

````python
        sas_logits, sa_logits = self.classifier(src_state, src_action, src_next_state, with_noise=False)
        sas_probs, sa_probs = F.softmax(sas_logits, -1), F.softmax(sa_logits, -1)
        sas_log_probs, sa_log_probs = torch.log(sas_probs + 1e-10), torch.log(sa_probs + 1e-10)
        reward_penalty = sas_log_probs[:, 1:] - sa_log_probs[:, 1:] - sas_log_probs[:, :1] + sa_log_probs[:,:1]
````

**Comment by OffDynamicsRL** [comment]
Hi, good catch! After checking our implementation and the official implementation by DARC authors (https://github.com/google-research/google-research/blob/master/darc/darc_agent.py), I think you are right. I have removed the following lines:
```
sas_probs, sa_probs = F.softmax(sas_logits, -1), F.softmax(sa_logits, -1)
```
These are unnecessary. Thank you for catching this!

**Comment by OffDynamicsRL** [comment]
BTW, we have also modified similar problems in other implemented algorithms. If you find any other problems, please do not hesitate to contact us!

**Comment by OffDynamicsRL** [comment]
I think the influence can be minor. I will be running experiments on some selected tasks to further check whether this affects the performance of our implemented DARC. I would post the results here as soon as possible

**Comment by OffDynamicsRL** [comment]
Hi, sorry for the late reply. I have completed some (comparatively large-scale) experiments using the implemented DARC algorithm on some selected environments, including 12 tasks from friction shift, gravity shift, kinematic shift and morphology shift. The results are shown below, where `DARC (previous)` denotes the implemented DARC algorithm with double softmax (i.e., extra softmax on the sa and sas logits), and `DARC (revise)` indicates that the additional softmax operator is removed. It can be seen that `DARC (previous)` beats `DARC (revise)` on some tasks while it can also underperform `DARC (revise)`. In some environments, their performance deviation can be minor. Considering the fact that the original DARC implementation does not involve the extra softmax operator, **it is recommended to use `DARC (revise)`**.

<img width="3053" height="1797" alt="Image" src="https://github.com/user-attachments/assets/979d09b8-ffa6-4266-be0f-156f6cdde3e1" />

To obtain the above results, I run `DARC (previous)` for 5 seeds, and `DARC (revise)` for 3 seeds. The logged data of DARC on the above tasks can be available upon request. Please let me know whether this addresses your issue or if you have any other problems. I appreciate your patience, and thank you again for catching the double softmax problem!

**Comment by stwerner97** [comment]
Thanks for the extensive experiments and interesting results! Yes, this addresses my concerns and I'll close this issue.
